internal/storage/block_storage.go

```diff
@@ -396,23 +381,161 @@ func (b *BlockStorage) CreateBlockCache(ctx context.Context) []*types.BlockIdent
 	return cache
 }
 
-// storeHash stores either a block or transaction hash.
-// TODO: store block hash in transaction hash value
-func (b *BlockStorage) storeHash(
+func (b *BlockStorage) storeBlockHash(
 	ctx context.Context,
 	transaction DatabaseTransaction,
-	hashKey []byte,
+	block *types.BlockIdentifier,
 ) error {
+	hashKey := getBlockHashKey(block)
 	exists, _, err := transaction.Get(ctx, hashKey)
 	if err != nil {
 		return err
 	}
 
 	if exists {
-		return ErrDuplicateKey
+		return fmt.Errorf("%w: duplicate block hash %s found", ErrDuplicateBlockHash, block.Hash)
 	}
 
 	return transaction.Set(ctx, hashKey, []byte(""))
 }
 
-// TODO: Wait until a transaction has a depth of X
+func (b *BlockStorage) storeTransactionHash(
+	ctx context.Context,
+	transaction DatabaseTransaction,
+	blockIdentifier *types.BlockIdentifier,
+	transactionIdentifier *types.TransactionIdentifier,
+) error {
+	hashKey := getTransactionHashKey(transactionIdentifier)
+	exists, val, err := transaction.Get(ctx, hashKey)
+	if err != nil {
+		return err
+	}
+
+	if !exists {
+		m := make(map[string]int64)
+		m[blockIdentifier.Hash] = blockIdentifier.Index
+		encodedResult, err := encode(m)
+		if err != nil {
+			return fmt.Errorf("%w: unable to encode transaction data", err)
+		}
+
+		return transaction.Set(ctx, hashKey, encodedResult)
+	}
+
+	var blocks map[string]int64
+	if err := decode(val, &blocks); err != nil {
+		return fmt.Errorf("%w: could not decode transaction hash contents", err)
+	}
+
+	if _, exists := blocks[blockIdentifier.Hash]; exists {
+		return fmt.Errorf(
+			"%w: duplicate transaction %s found in block %s:%d",
```

Comment:
Getting a little confused here since we are retrieving the block info from the blocks map. Not sure we have duplicate transaction in the error message here

Reply:
We detect duplicate transactions (this would be 2 transaction hashes in the same block) if we try to add the same block hash to the same transaction block map more than once.

We very purposely do not error if there are 2 identical transaction hashes in different blocks (#57  and #53).

Reply:
Ah this makes sense now